### PR TITLE
CUMULUS-1152: compare file holdings in Cumulus and CMR

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-1152: compare file holdings in Cumulus and CMR](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1152)

## Changes

* compare file holdings for granules which are in both Cumulus and CMR

## PR Checklist

- [x ] Update CHANGELOG
- [x ] Unit tests
- [x ] Adhoc testing
- [x] Integration tests

### CODE REVIEW

File: packages/api/lambdas/create-reconciliation-report.js [hunk 1]
```diff
@@ -242,34 +338,47 @@ exports.reconciliationReportForGranules = reconciliationReportForGranules;
...
+    okCount: 0,
     onlyInCumulus: [],
     onlyInCmr: []
   };
-  granuleReports.forEach((granuleReport) => {
-    granulesInCumulusCmr.okGranuleCount += granuleReport.okGranulesInDb.length;
-    granulesInCumulusCmr.onlyInCumulus = granulesInCumulusCmr
-      .onlyInCumulus.concat(granuleReport.onlyInCumulus);
-    granulesInCumulusCmr.onlyInCmr = granulesInCumulusCmr.onlyInCmr.concat(granuleReport.onlyInCmr);
+  granuleAndFilesReports.forEach((granuleFileReport) => {
```
Comment (markdboyd) [inline packages/api/lambdas/create-reconciliation-report.js]: Seems like a `.reduce()` operation might be appropriate here and more functional, but it's not strictly necessary

File: packages/api/tests/lambdas/create-reconciliation-report.js [hunk 2]
```diff
@@ -542,20 +546,120 @@ test.serial('reconciliationReportForGranules reports discrepancy of granule hold
 
   await new models.Granule().create(matchingGrans.concat(extraDbGrans));
 
-  const granulesReport = await reconciliationReportForGranules(collectionId);
-
-  const expectedOkGranulesInDb = sortBy(matchingGrans, ['granuleId']).map((gran) =>
```
Comment (markdboyd) [inline packages/api/tests/lambdas/create-reconciliation-report.js]: Why remove the granule ID comparisons?

File: example/spec/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 3]
```diff
@@ -221,7 +265,7 @@ describe('When there are granule differences and granule reconciliation is run',
 
   it('generates a report showing number of collections that are in both Cumulus and CMR', () => {
     // MOD09GQ___006 is in both Cumulus and CMR
-    expect(report.collectionsInCumulusCmr.okCollectionCount).toBeGreaterThan(0);
+    expect(report.collectionsInCumulusCmr.okCount).toBeGreaterThanOrEqual(1);
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply example/spec/createReconciliationReport/CreateReconciliationReportSpec.js]: There might be other collections people added to their stacks and CMR.  So we only check there should be at least one collection in both Cumulus and CMR.

File: packages/api/tests/lambdas/create-reconciliation-report.js (hunk 2 above)
Comment (jennyhliu) [reply packages/api/tests/lambdas/create-reconciliation-report.js]: The function no longer returns the `ok` granules.  If the granule is in both Cumulus and CMR, its files are compared   right away.

File: example/spec/createReconciliationReport/CreateReconciliationReportSpec.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (markdboyd) [reply example/spec/createReconciliationReport/CreateReconciliationReportSpec.js]: Ah, of course 😅 . Thanks for the explanation

File: packages/api/lambdas/create-reconciliation-report.js (hunk 1 above)
Comment (jennyhliu) [reply packages/api/lambdas/create-reconciliation-report.js]: Good suggestion.  I updated the code to use `reduce()`

File: packages/api/lambdas/create-reconciliation-report.js [hunk 4]
```diff
@@ -146,14 +149,93 @@ async function reconciliationReportForCollections() {
...
+        });
+
+        if (accessUrl && relatedUrl.URL === accessUrl.URL) {
+          okCount += 1;
+        }
+        else if (cmrGetDataTypes.includes(relatedUrl.Type)) {
+          // ignore any URL which is not for getting data
+          // some files should not be in CMR such as private files
+          onlyInCmr.push({
+            URL: relatedUrl.URL, Type: relatedUrl.Type, GranuleUR: granuleInCmr.GranuleUR
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply packages/api/lambdas/create-reconciliation-report.js]: updated

File: packages/cmrjs/cmr-utils.js [hunk 5]
```diff
@@ -306,13 +306,46 @@ function getS3CredentialsObject(s3CredsUrl) {
...
+/**
+ * Construct online access url for a given file.
+ *
+ * @param {Object} params - input parameters
+ * @param {Object} params.file - file object
+ * @param {string} params.distEndpoint - distribution endpoint from config
+ * @param {BucketsConfig} params.buckets -  stack BucketConfig instance
+ * @returns {Object} online access url object
+ */
+function constructOnlineAccessUrl({
```
Comment (ifestus) [inline packages/cmrjs/cmr-utils.js]: I'm a fan of this  change :+1:

File: packages/api/lambdas/create-reconciliation-report.js [hunk 6]
```diff
@@ -147,13 +150,95 @@ async function reconciliationReportForCollections() {
...
+          GranuleUR: granuleInCmr.GranuleUR
+        });
+      }
+    }
+  });
+
+  // any remaining database items to the report
+  Object.keys(granuleFiles).forEach((fileName) => {
+    // private file only in database, it's ok
+    if (bucketsConfig.key(granuleFiles[fileName].bucket)
```
Comment (markdboyd) [inline packages/api/lambdas/create-reconciliation-report.js]: So a private file that is not posted to CMR and only tracked by the Cumulus database is not considered "only in Cumulus" for the purpose of the reconciliation report?

Comment (jennyhliu) [reply packages/api/lambdas/create-reconciliation-report.js]: The private file is not supposed to be in CMR granule, and it's in Cumulus only, and shouldn't be reported as a discrepancy

**Pull request merged**